### ISSUE TITLE
no-unused-variable configuration request

### ISSUE THREAD

**Issue by scriby** [palantir/tslint#191]
As the no-unused-variable is implemented now, I find it has too many false positives to really be useful. However, if I could just enable a portion of it that would check only local variables, that would be really useful.

**Comment by JoshuaKGoldberg** [comment]
Sorry if bumping this is against the rules - I've come across the same problem.

```typescript
    class Test {
        private done: boolean;

        constructor() {
            (function (): void {
                this.doPrivateThing();
                (<Test>this).doPrivateThing();
            }).bind(this)();
        }

        private doPrivateThing(): void {
            this.done = false;
        }
    }
```

...gives the following error:

    >> test.ts[11, 13]: unused variable: 'doPrivateThing'

It looks like tslint isn't checking inside the bound function. It's understandable that function.bind erases some information, but should `<Test>this.doPrivateThing` let it know that the variable's being used?

**Comment by sundaysen** [comment]
I have a similar problem maybe?
A variable used for type declaration will trigger the no-unused-variable rule.

```
import Foo = this.that.Foo;

module some.module.blah {
 export  class bar {
    private foo: Foo;
  }
}
```

returns the error  `unused variable: 'Foo`

**Comment by nikklassen** [comment]
The use of an import alias is also not detected when it's used with extends

```ts
import DateTimeOpts = Intl.DateTimeFormatOptions;

interface MyDateTimeOpts extends DateTimeOpts {
    timezoneOffset: number;
}

let opts: MyDateTimeOpts;
console.log(opts.timezoneOffset - 1);
```

**Comment by adidahiya** [comment]
@SenHeng, @nikklassen: those look like related issues; I've filed a new ticket for those false positives: #613.

as for the other issues in this thread, I think we could address them by adding a new option that makes the `no-unused-variable` rule skip private class members: #614. I don't want to do anything much more complex / involved than that; I really don't want to encourage any coding style that references private members with dynamic access as suggested in some of the examples above.

**Comment by adidahiya** [comment]
closing this for now. I can reopen if someone wants to discuss further... or feel free to file a new, more specific issue

**Comment by JoshuaKGoldberg** [comment]
> I think we could address them by adding a new option that makes the `no-unused-variable` rule skip private class members

Forgive me if I'm misunderstanding, but does that not solve a different problem? Caring whether private variables are unused is different from correctly reporting whether they're unused, yes?

For further clarification, in the below example, `waste` should be flagged as unused, but `done` should not be.

```typescript
class Test {
    private waste: boolean;
    private done: boolean;

    constructor() {
        (function (): void {
            this.doPrivateThing();
            (<Test>this).doPrivateThing();
        }).bind(this)();
    }

    private doPrivateThing(): void {
        this.done = false;
    }
}
```

**Comment by adidahiya** [comment]
ok, I think I see what you mean with the difference between `waste` and `done` there... but I don't see a straightforward approach that would correctly mark `waste` as unused while _not also marking `doPrivateThing()` as unused_. any ideas here?

basically the option I've suggested in #614 is a less strict version of the `no-unused-variable` rule for users who want to use dynamic access of private members/methods _at all_ in their codebase.

**Comment by JoshuaKGoldberg** [comment]
Unfortunately I'm not familiar with the tslint codebase, so that brings up a question. Should the following line above should mark `doPrivateThing` as used?

```typescript
(<Test>this).doPrivateThing();
```

...which brings up another question: does tslint know the anonymous function is being executed?

```typescript
    constructor() {
        (function (): void {
            (<Test>this).doPrivateThing();
        }).bind(this)();
    }
```

**Comment by adidahiya** [comment]
No, we don't do much control flow analysis to know things like "is this anonymous function being executed". I'm hesitant to do that unless we have really good reasons to do so (it complicates the codebase).

`(<Test>this).doPrivateThing()` is the kind of dynamic access I'm referring to, and I don't even think the compiler recognizes that as a static reference to the private class method (@JKillian -- can you confirm please?). It mainly has to do with `Function.prototype.bind`, which erases some type information. So if the compiler doesn't count it as a reference, tslint will not do so either.

**Comment by JKillian** [comment]
@JoshuaKGoldberg 
There are [currently proposals](https://github.com/Microsoft/TypeScript/issues/3694) on supporting typing for `this`, but until then, TSLint won't be able to recognize that `this.doPrivateThing();` from your example refers to an instance of `Test` and is using its method. And you can see why this would be tricky: the `this` in the function could be anything. (Just as a note, when in a class method, `this` is typed.)

However, I _think_ `(<Test>this).doPrivateThing()` should be recognized as a reference to the method by TypeScript and thus and should not be marked as unused by TSLint. Is TSLint marking the method as unused? If so I'll look into it.

**Comment by JKillian** [comment]
@JoshuaKGoldberg If I run TSLint on your example, I get an unused variable warning for `waste` but not for anything else. This is how it should work, right? :smiley: Are you getting differing results?

**Comment by JoshuaKGoldberg** [comment]
> However, I think  (<Test>this).doPrivateThing()  should be recognized as a reference to the method by TypeScript and thus and should not be marked as unused by TSLint. Is TSLint marking the method as unused? If so I'll look into it.

This was the behavior in July, but I just ran it locally and seems to be working. Thanks for taking the time to look into it, @adidahiya and @JKillian ! Everything seems to be fixed :)

**Comment by JKillian** [comment]
No problem! Always nice when bugs disappear haha
